### PR TITLE
Return list column in dm_examine_constraints()

### LINKED ISSUE #239: Prepare dm_examine_constaints() for compound keys
Only for the interface, let's not change internals for now.

Should return `vctrs::list_of()`, create with:

```r
new_keys <- function(x) {
  vctrs::new_list_of(x, character())
}
```

- [x] `dm_examine_constraints()` should return a list in the `column` column. Can we keep `dm_check_constraints()` as is?
- [x] `dm_get_all_?ks()`
- [x] `dm_enum_?k_candidates()`
- [x] `enum_?k_candidates()`
- [x] `dm_get_pk()` et al.: Adapt documentation to describe that multiple key columns may be possible, perhaps a short vignette?

### PR DESCRIPTION
**Pull request by krlmlr**
Closes #239.

@TSchiefer: Can you please pick up from here?

### CODE REVIEW

**Comment by TSchiefer** [comment]
@krlmlr My local R CMD check runs nicely and I think I addressed the points in #239. I cannot request a review, since you started the PR, but could you have a look?

File: R/foreign-keys.R [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
+#' @description `dm_add_fk()` marks the specified column as the foreign key of table `table` with
+#' respect to the primary key of table `ref_table`.
+#' If `check == TRUE`, then it will first check if the values in column `column` are a subset
+#' of the values of the primary key in table `ref_table`.
+#'
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
+#' therefore the syntax of this function will slightly change.
+#'
 #' @inheritParams dm_add_pk
 #' @param column The column of `table` which is to become the foreign key column and
```
Comment (krlmlr) [inline R/foreign-keys.R]: Here and everywhere?

```suggestion
#' @param columns The column of `table` which is to become the foreign key column and
```

File: R/foreign-keys.R [hunk 2]
```diff
@@ -1,5 +1,13 @@
 #' Add a reference from one table of a [`dm`] to another
 #'
+#' @description `dm_add_fk()` marks the specified column as the foreign key of table `table` with
+#' respect to the primary key of table `ref_table`.
+#' If `check == TRUE`, then it will first check if the values in column `column` are a subset
+#' of the values of the primary key in table `ref_table`.
+#'
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
```
Comment (krlmlr) [inline R/foreign-keys.R]:
```suggestion
#' @section Compound keys:
#'
#' Currently, keys consisting of more than one column are not supported.
#' [This feature](https://github.com/krlmlr/dm/issues/3) is planned for dm 0.2.0.
#' The syntax of this function will be extended but remain compatible
#' with current semantics.
```

File: R/foreign-keys.R [hunk 3]
```diff
@@ -1,5 +1,13 @@
...
+#' therefore the syntax of this function will slightly change.
```
Comment (krlmlr) [inline R/foreign-keys.R]:
```suggestion
```

File: R/foreign-keys.R [hunk 4]
```diff
@@ -85,6 +93,13 @@ dm_has_fk_impl <- function(dm, table_name, ref_table_name) {
 
 #' Retrieve the name of the column marked as a foreign key, pointing from one table of a [`dm`] to another table.
 #'
+#' @description `dm_get_fk()` returns the name of the
+#' column marked as foreign key of table `table` with respect to table `ref_table` within a [`dm`] object.
+#' If no foreign key is set between the tables, an empty character vector is returned.
+#'
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
+#' therefore the result of this function will slightly change.
```
Comment (krlmlr) [inline R/foreign-keys.R]: see above, using

```suggestion
#' therefore the function may return vectors of length greater than one in the future.
```

File: R/foreign-keys.R [hunk 5]
```diff
@@ -131,7 +149,8 @@ dm_get_fk_impl <- function(dm, table_name, ref_table_name) {
 #' @export
 dm_get_all_fks <- nse(function(dm) {
   check_not_zoomed(dm)
-  dm_get_all_fks_impl(dm)
+  dm_get_all_fks_impl(dm) %>%
+    mutate(child_fk_col = new_keys(child_fk_col))
```
Comment (krlmlr) [inline R/foreign-keys.R]:
```suggestion
    mutate(child_fk_cols = new_keys(child_fk_cols))
```

File: R/foreign-keys.R [hunk 6]
```diff
@@ -115,6 +130,9 @@ dm_get_fk_impl <- function(dm, table_name, ref_table_name) {
 #'
 #' @description Get a summary of all foreign key relations in a [`dm`]
 #'
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
+#' therefore the result of this function will slightly change.
```
Comment (krlmlr) [inline R/foreign-keys.R]: See above, using

```suggestion
#' therefore the `child_fk_cols` column may contain vectors of length greater than one.
```

File: R/foreign-keys.R [hunk 7]
```diff
@@ -144,6 +163,8 @@ dm_get_all_fks_impl <- function(dm) {
 #'
 #' @description This function can remove either one reference between two tables, or all references at once, if argument `column = NULL`.
 #' All arguments may be provided quoted or unquoted.
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
+#' therefore the syntax of this function will slightly change.
```
Comment (krlmlr) [inline R/foreign-keys.R]: Same syntax as `dm_add_fk()`, perhaps document on the same page?

File: R/foreign-keys.R [hunk 8]
```diff
@@ -144,6 +163,8 @@ dm_get_all_fks_impl <- function(dm) {
...
 #'
 #' @inheritParams dm_add_fk
 #' @param column The column of `table` that should no longer be referencing the primary key of `ref_table`.
```
Comment (krlmlr) [inline R/foreign-keys.R]:
```suggestion
#' @param columns The column of `table` that should no longer be referencing the primary key of `ref_table`.
```

File: R/foreign-keys.R [hunk 9]
```diff
@@ -260,7 +281,9 @@ dm_enum_fk_candidates <- nse(function(dm, table, ref_table) {
   ref_tbl <- tbl(dm, ref_table_name)
   tbl <- tbl(dm, table_name)
 
-  enum_fk_candidates_impl(table_name, tbl, ref_table_name, ref_tbl, ref_tbl_pk)
+  enum_fk_candidates_impl(table_name, tbl, ref_table_name, ref_tbl, ref_tbl_pk) %>%
+    rename(columns = column) %>%
```
Comment (krlmlr) [inline R/foreign-keys.R]: Please adapt the data description in the help.

File: R/key-helpers.R [hunk 10]
```diff
@@ -8,7 +8,7 @@
 #'   \describe{
```
Comment (krlmlr) [inline R/key-helpers.R]: Each function changed here can contain a "Compound keys" section.

File: R/primary-keys.R [hunk 11]
```diff
@@ -8,6 +8,9 @@
 #' If `force == TRUE`, the function will replace an already
 #' set key.
 #'
+#' FYI: Mind, that in one of the upcoming updates, compound keys will be supported, and
+#' therefore the syntax for this function will slightly change.
+#'
 #' @param dm A `dm` object.
 #' @param table A table in the `dm`.
 #' @param column A column of that table.
```
Comment (krlmlr) [inline R/primary-keys.R]:
```suggestion
#' @param columns A table column, unquoted.
```

File: R/foreign-keys.R (hunk 7 above)
Comment (TSchiefer) [reply R/foreign-keys.R]: Ok, and for symmetry reasons I also merged `dm_add_pk()` and `dm_rm_pk()`

**Comment by TSchiefer** [comment]
- [x] Release CI test passing
- [ ] R 3.3 CI test fails, apart from the fnd-role topic there is a problem with an error class in the tests, even though the error msg is as expected: https://travis-ci.org/krlmlr/dm/jobs/640888046?utm_medium=notification&utm_source=email

I think it should probably be alright though.

**Comment by krlmlr** [comment]
Thanks! Tests seem to be good now.

**Pull request merged**